### DISCUSSION TITLE
LDAP login tries to search in a hardcoded search base

### DISCUSSION THREAD

**Discussion by viharm** [firefly-iii/firefly-iii#5533] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

With LDAP authentication configured, I am unable to login. Not even for the first time.

### Expected behaviour

Be able to login.

### Debug information

I tried installing without LDAP and was able to register and login successfully.

I then tried various configurations

## With a bind user

The logs show that the base dn being searched is `dc=local,dc=com`, even though I've explicitly configured `LDAP_BASEDN`
Log message by _OpenLDAP_  from _syslog_ of host machine.
```
slapd[31545]: conn=9395241 op=1 SRCH base="dc=local,dc=com" scope=2 deref=0 filter="(&(objectClass=top)(objectClass=person)(objectClass=organizationalPerson)(objectClass=inetOrgPerson)(uid=USERNAME))"
```
I have replaced my aactual username with `USERNAME` in the above log. The other generic items like the search based are actually from the log.

## Without a bind user

The log shows a connection from a _Docker_ IP trying to bind with a user `cn=user,dc=local,dc=com`.

Log message by _OpenLDAP_  from _syslog_ of host machine.
```
slapd[31545]: conn=9423095 op=0 BIND dn="cn=user,dc=local,dc=com" method=128
```

_Docker_ container log
```
[2022-01-12 19:05:29] production.WARNING: Login attempt failed.  
```

### Steps to reproduce

Create a _Docker_ container with the options shown below.

```ini
-e TRUSTED_PROXIES=**
-e APP_DEBUG=true
-e APP_LOG_LEVEL=debug
-e LOG_CHANNEL=stack
-e AUDIT_LOG_LEVEL=info
-e DEFAULT_LANGUAGE=en_GB
-e TZ="Europe/London"
-e APP_KEY=32characterkey
-e APP_NAME=InstallLabel
-e APP_ENV=production
-e APP_URL=https://domain.tld
-e SITE_OWNER=myadmin
-e DB_CONNECTION=pgsql
-e DB_HOST=10.0.0.3
-e DB_PORT=5432
-e DB_DATABASE=databasename
-e DB_USERNAME=databaseuser
-e DB_PASSWORD=databasepass
-e AUTHENTICATION_GUARD=ldap
-e LDAP_DIALECT=OpenLDAP
-e LDAP_HOST=10.0.0.3
-e LDAP_PORT=389
-e LDAP_SSL=false
-e LDAP_TLS=false
-e LDAP_BASEDN="ou=Users,dc=domain,dc=tld"
-e LDAP_AUTH_FIELD=uid
-e LDAP_EXTRA_FILTER="(objectClass=posixAccount)"
-e MAIL_MAILER=smtp
-e MAIL_HOST=smtpmailserver.net
-e MAIL_PORT=465
-e MAIL_ENCRYPTION=ssl
-e MAIL_USERNAME=mailuser
-e MAIL_PASSWORD=mailpassword
-e MAIL_FROM="applabel@domain.tld"
-e SEND_REGISTRATION_MAIL=true
-e SEND_ERROR_MESSAGE=true
-e SEND_LOGIN_NEW_IP_WARNING=true
```

### Additional info

I'm trying to setup _Firefly III_ with the following environment
  - _Firefly III_ image 5.6.10
  - _PostgreSQL_ database 11.14
  - _Docker_ 20.10.11
  - _OpenLDAP_ 2.4.47
  - _Nginx_ 1.14.2 (reverse proxy)

All installed on the same physical machine.

I'm using the following environment variables for my docker command

**Comment by tjmv** [comment]
https://github.com/firefly-iii/firefly-iii/blob/7afe9fac0af0ceac992064a6d3ab35a912122b6b/.env.example#L195
https://github.com/firefly-iii/firefly-iii/blob/7afe9fac0af0ceac992064a6d3ab35a912122b6b/config/ldap.php#L60

`LDAP_BASE_DN`. Notice the underscore.

**Reply by viharm** [reply]
So correcting the env var name to `LDAP_BASE_DN` (with the underscore) worked. Thanks @tjmv 

However, it requires a fixed bind user to work. As soon as I remove the bind user, auth fails. Is the bind user a requirement for the underlying LDAP authentication library?

**Reply by JC5** [reply]
It needs a bind user to search the LDAP directory for your authentication details, group, etc.

**Comment by JC5** [comment]
LDAP support has some issues, but it's low on my priority list at the moment.

**Reply by gkl1368** [reply]
I also meet this problem on 5.6.16, the ldap auth only worked with no group filter in base dn like "dc=domain,dc=local". If i add group filter like "cn=firefly-group,cn=groups,dc=domain,dc=local", it always show " Login attempt failed. "

**Comment by gramakri** [comment]
@viharm The object classes are hardcoded in the LDAP module . See my comment at https://github.com/firefly-iii/firefly-iii/issues/5133#issuecomment-1013561081 . They come from https://github.com/DirectoryTree/LdapRecord/blob/master/src/Models/OpenLDAP/User.php#L33

**Reply by viharm** [reply]
Understood, thanks. Although this wasn't a problem for me (since I have a vanilla-like _OpenLDAP_ setup), I was thinking of experimenting with a separate OU with different object classes. You saved me some long frustrating hours :-)
